dev/com.ibm.ws.concurrent.mp.cdi/src/com/ibm/ws/concurrent/mp/cdi/ConcurrencyCDIExtension.java

```diff
@@ -36,13 +56,129 @@
+        String name = nameAnno == null ? member.getDeclaringClass().getTypeName() + "." + member.getName() : nameAnno.value();
         if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled())
-            Tr.debug(tc, "Registering MP Concurrency producers");
-        AnnotatedType<ManagedExecutorProducer> execType = bm.createAnnotatedType(ManagedExecutorProducer.class);
-        event.addAnnotatedType(execType, ManagedExecutorProducer.class.getName());
+            Tr.debug(tc, "InjectionPoint " + name + " in app " + appName + " has config " + config);
+        if (nameAnno == null)
+            event.configureInjectionPoint().addQualifiers(createNamedInstance(name));
+        ManagedExecutorConfig prevoiusConfig = appBeanMap.putIfAbsent(name, config);
+
+        // If 2 or more InjectionPoints define @NamedInstance("X") @ManagedExecutorConfig it is an error
+        if (prevoiusConfig != null) {
+            String msg = "ERROR: Found existing bean with name=" + name; // TODO NLS message
+            Tr.error(tc, msg);
+            deploymentErrors.add(new Throwable(msg));
+        }
+    }
+
+    private CDIService getCDIService() {
+        if (cdiSvc != null)
+            return cdiSvc;
+
+        Bundle bundle = FrameworkUtil.getBundle(CDIService.class);
+        return cdiSvc = AccessController.doPrivileged((PrivilegedAction<CDIService>) () -> {
+            BundleContext bCtx = bundle.getBundleContext();
+            ServiceReference<CDIService> svcRef = bCtx.getServiceReference(CDIService.class);
+            return svcRef == null ? null : bCtx.getService(svcRef);
+        });
     }
 
+    public void registerBeans(@Observes AfterBeanDiscovery event, BeanManager beanManager) {
+        CDIService cdiSvc = getCDIService();
+        if (!cdiSvc.isCurrentModuleCDIEnabled())
+            return;
+
+        String appName = cdiSvc.getCurrentApplicationContextID();
+        if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled())
+            Tr.debug(tc, "Register beans for app " + appName);
+
+        // TODO: Right now this is coded to give 1 ManagedExecutor instance for all @Default injection points
+        // wait to see how discussion on this PR ends up: https://github.com/eclipse/microprofile-concurrency/pull/54
+        // Always register bean for default instance(s)
+        event.addBean(new ManagedExecutorBean());
+
+        // Register 1 bean per un-named config, and 1 bean per unique NamedInstance
+        for (Entry<String, ManagedExecutorConfig> e : beanMap.get(appName).entrySet()) {
+            if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled())
+                Tr.debug(tc, "Add bean for name=" + e.getKey());
+            event.addBean(new ManagedExecutorBean(e.getKey(), e.getValue()));
+        }
+    }
+
+    public void registerErrors(@Observes AfterDeploymentValidation event) {
+        deploymentErrors.forEach(event::addDeploymentProblem);
```

Comment:
deploymentErrors does not ever appear to be cleared.  If someone makes multiple attempts iterating over installing an application, will it keep accumulating/retaining errors from previous attempts?